src/cache.cpp

```diff
@@ -746,6 +767,101 @@ bool StatCache::DelSymlink(const char* key, AutoLock::Type locktype)
     return true;
 }
 
+// [NOTE]
+// Need to lock StatCache::stat_cache_lock before calling this method.
+//
+bool StatCache::AddNotruncateCache(const std::string& key)
+{
+    std::string parentdir = mydirname(key);
+    std::string filename  = mybasename(key);
+
+    if(!parentdir.empty() && !filename.empty()){
+        return false;
+    }
+    parentdir += '/';       // directory path must be '/' termination.
+
+    notruncate_dir_map_t::iterator iter = notruncate_file_cache.find(parentdir);
+    if(iter == notruncate_file_cache.end()){
+        // add new list
+        notruncate_filelist_t list;
+        list.push_back(filename);
+        notruncate_file_cache[parentdir] = list;
+    }else{
+        // add filename to existed list
+        notruncate_filelist_t& filelist = iter->second;
+        notruncate_filelist_t::const_iterator fiter = std::find(filelist.begin(), filelist.end(), filename);
+        if(fiter == filelist.end()){
+            filelist.push_back(filename);
+        }
+    }
+    return true;
+}
+
+// [NOTE]
+// Need to lock StatCache::stat_cache_lock before calling this method.
+//
+bool StatCache::DelNotruncateCache(const std::string& key)
+{
+    std::string parentdir = mydirname(key);
+    std::string filename  = mybasename(key);
+
+    if(!parentdir.empty() && !filename.empty()){
+        return false;
+    }
+    parentdir += '/';       // directory path must be '/' termination.
+
+    notruncate_dir_map_t::iterator iter = notruncate_file_cache.find(parentdir);
+    if(iter != notruncate_file_cache.end()){
+        // found directory in map
+        notruncate_filelist_t& filelist = iter->second;
+        notruncate_filelist_t::iterator fiter = std::find(filelist.begin(), filelist.end(), filename);
+        if(fiter != filelist.end()){
+            // found filename in directory file list
+            filelist.erase(fiter);
+            if(filelist.empty()){
+                notruncate_file_cache.erase(parentdir);
+            }
+        }
+    }
+    return true;
+}
+
+// [Background]
+// When s3fs creates a new file, the file does not exist until the file contents
+// are uploaded.(because it doesn't create a 0 byte file)
+// From the time this file is created(opened) until it is uploaded(flush), it
+// will have a Stat cache with the No truncate flag added.
+// This avoids file not existing errors in operations such as chmod and utimens
+// that occur in the short period before file upload.
+// Besides this, we also need to support readdir(list_bucket), this method is
+// called to maintain the cache for readdir and return its value.
```

Comment:
I'm confused about how this happens -- while the file is uploading, the stat object must continue to exist.  Why doesn't readdir find this entry?

Reply:
In `readdir` is processing , `list_bucket()` is called.
`list_bucket` always calls `ListBucketRequest` to list the objects that currently exist.
(This is necessary in order to detect objects manipulated from other than s3fs in the S3 bucket.)
After calling `list_bucket`, `readdir_multi_head` is called to get information about each object.
The Stat cache will be accessed for objects listed in this `readdir_multi_head`.

The `readdir` process is as described above, and while a new file is being created(from `create` to `flush`), the file does not exist in the results of `ListBucketRequest`.
So this PR, if there is a temporary Stat cache for a newly created file, the file name is added to the list of `ListBucketRequest`.
With this addition, newly created files can also be read using readdir.